### ISSUE TITLE
Kibana app doesn't connect  to API

### ISSUE THREAD

**Issue by antgel** [wazuh/wazuh-kibana-app#1628]
| Wazuh | Elastic | Rev |
| ----- | ------- | --- |
| 3.9.3   | 7.2.0     |  |

**Description**

Can't connect Kibana app to API.

**Steps to reproduce**

1. Set up entire stack, it's a fresh install.
2. Curl works:
```
curl -k -u MUNGED:MUNGED 'https://localhost:55000?pretty'
{
   "error": 0,
   "data": {
      "msg": "Welcome to Wazuh HIDS API",
      "api_version": "v3.9.3",
      "hostname": "ip-172-31-67-156",
      "timestamp": "Wed Jul 24 2019 14:59:19 GMT+0000 (Coordinated Universal Time)"
   }
}
```

Curl also works from my local workstation to the server, so I know the network is okay. But from Kibana... (see screenshot)

**Screenshots**

![image](https://user-images.githubusercontent.com/295571/61802691-93f22680-ae39-11e9-858c-8be96a2151e6.png)

**Additional context**

Happy to provide more information if you tell me what you need. I wish I knew where to find the log for the 500 Internal Server Error.

**Comment by juankaromo** [comment]
Hi @antgel 

I would like to see the content of that error 500 message that you got from the checkAPI call in order to know what's the message you got from that failed call.

Open the browser console. Go to the Network tab, and then locate the XHR tab. Now try to register an API, and if you're still getting the error, you'll see a message on that tab, colored in red. Click on it and then, go to the Response tab. Please, send us the content to see what's the cause of your problem.

I'll attach a screenshot to help you find the tab on Google Chrome.

![image](https://user-images.githubusercontent.com/21195730/61851471-621fa500-aeb7-11e9-87f6-88e7c1aaa5b1.png)

Please, paste here the output.

Best regards,

**Comment by antgel** [comment]
Wow, thanks for such a helpful response. The error is: `{"message":"3005 - Wrong protocol being used to connect to the Wazuh API","code":3005,"statusCode":500}`. I have searched this message online and found little of use. Using wazuh-api 3.9.3, wazuh-manager 3.9.3, Kibana 7.2.0, Kibana plugin 3.9.3. What can I check next?

**Comment by juankaromo** [comment]
Hi @antgel 

I think that the response the API is giving to you is caused because the API is configured with `http` but your request is being performed with `https`. Please, could you try again the same request but using `http` protocol in the URL?

Best regards,

**Comment by antgel** [comment]
Hi @juankaromo 

I tried both, and it was the same error. I followed the onboarding wizard, My config (after the install wizard) looks like:

```
var config = {};

// Basic configuration

// Path
config.ossec_path = "/var/ossec";
// The host to bind the API to.
config.host = "0.0.0.0";
// TCP Port used by the API.
config.port = "55000";
// Use HTTP protocol over TLS/SSL. Values: yes, no.
config.https = "yes";
// Use HTTP authentication. Values: yes, no.
config.basic_auth = "yes";
//In case the API run behind a proxy server, turn to "yes" this feature. Values: yes, no.
config.BehindProxyServer = "no";

// HTTPS Certificates
//config.https_key = "configuration/ssl/server.key"
//config.https_cert = "configuration/ssl/server.crt"
//config.https_use_ca = "no"
//config.https_ca = "configuration/ssl/ca.crt"
```

Another weird smell is that when I set `config.logs = "debug";' in `/var/ossec/api/configuration/config.js` then `systemctl restart wazuh-api`, I don't seem to get any extra log information in `/var/ossec/logs/ossec.log`. This makes me think that my configuration isn't being read, but perhaps I'm paranoid. Where can I go next with this?

**Comment by juankaromo** [comment]
Hi @antgel 

Are you in a single host or in a distributed architecture? Your configuration looks fine but reviewing the error you pointed out, I realized is possible that you are typing a wrong URL o for some reason you have not access to the typed URL.

I look forward to hearing from you.

Regards
Could you test if you have access from the browser to the API? Please type in the browser `https://<api_url>:55000` and check if you can access it.

On the other hand, have you configured a firewall that might be avoiding the connection?

**Comment by antgel** [comment]
Hi @juankaromo , you hit the nail on the head, I'll explain below:

My server is in AWS EC2. I originally had the AWS security group to allow access from my office public IP to port 55000. I assumed this was fine because when I went to `https://<api_url>:55000` in my browser, I did see the basic auth dialog, but as you know, I couldn't connect from Kibana.

Yet, when I open up 55000 fully, it works! If the correct IP address in the security group *isn't* my office public IP, what should it be?

**Comment by juankaromo** [comment]
Hi @antgel 

Assuming that you are in a single-host server, you could try attempt to connect with the API from Kibana app through the loopback IP address and establishing a local connection.

So, could you try to connect typing `https://127.0.0.1` or `https://localhost` on the app?

Tell me the results.

Regards,

**Comment by antgel** [comment]
`https://localhost` worked! Sorry for wasting your time, I don't know if this was meant to be somehow obvious or if I missed it in the docs. Really appreciate it.

**Comment by tbhaxor** [comment]
Its not working in my case

```
root@opendistro:~# curl https://localhost:55000
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
root@opendistro:~# curl http://localhost:55000
401 Unauthorizedroot@opendistro:~# 
```

**Comment by Yashvendra** [comment]
> Its not working in my case
> 
> ```
> root@opendistro:~# curl https://localhost:55000
> curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
> root@opendistro:~# curl http://localhost:55000
> 401 Unauthorizedroot@opendistro:~# 
> ```

It is showing unauthorised because you haven't passed credentials in the curl request which are required to access the api. If you want to remove auth you can do that by setting `config.basic_auth` to `no` in `/var/ossec/api/configuration/config.js`. Then do the above curl request, it should work just fine.

Also, you will have to remove `user` and `password` fields from `/usr/share/kibana/optimize/wazuh/config/wazuh.yml` in order for kibana to work with Wazuh API.

**Comment by Yashvendra** [comment]
Hi @juankaromo, 

My kibana is not able to communicate with Wazuh-API.
This is my `/var/ossec/api/configuration/config.js`
<img width="555" alt="wazuh3" src="https://user-images.githubusercontent.com/23001311/88160401-351de380-cc2c-11ea-8540-5f5d1441647e.PNG">

And this is my `/usr/share/kibana/optimize/wazuh/config/wazuh.yml`
<img width="187" alt="wazuh5" src="https://user-images.githubusercontent.com/23001311/88160662-96de4d80-cc2c-11ea-974a-545e7a36f77d.PNG">

When I do curl request (for both localhost/0.0.0.0), it says
![wazuh4](https://user-images.githubusercontent.com/23001311/88160738-b7a6a300-cc2c-11ea-8134-d04e682217cf.PNG)

But when I check connection from kibana, its not able to communicate with Wazuh-API.
<img width="524" alt="wazuh1" src="https://user-images.githubusercontent.com/23001311/88160813-d1e08100-cc2c-11ea-9c91-5df4b0423942.PNG">
<img width="525" alt="wazuh2" src="https://user-images.githubusercontent.com/23001311/88160824-d442db00-cc2c-11ea-9d2c-040f06d67f1d.PNG">

Can anyone share his thoughts on this?

**Comment by Joanes04** [comment]
Hi @Yashvendra ,

It seems that you do not have a well-defined `/usr/share/kibana/optimize/wazuh/config/wazuh.yml` file. The correct definition of a host is as follows: 

```
hosts:
  - default:
     url: https://localhost
     port: 55000
     user: foo
     password: bar

```
As you can see, you are missing the definition of the `user` and the `password`.

I hope I helped you

Regards,
Alberto

**Comment by Yashvendra** [comment]
Hi @Joanes04, thanks for replying.

However, I didn't include `user` and `password` because haven't set authentication for accessing Wazuh-API. If you look at `/var/ossec/api/configuration/config.js` (screenshot attached in the previous comment), I've set `config.basic_auth` to `no`. Therefore, I think I don't need to supply user and password in `wazuh.yml`.

Still, I followed what you said and supplied the creds. And this is the response I am getting now:
<img width="523" alt="wazuh6" src="https://user-images.githubusercontent.com/23001311/88166520-00faf080-cc35-11ea-9cce-ee06f600fc40.PNG">
<img width="664" alt="wazuh7" src="https://user-images.githubusercontent.com/23001311/88166526-035d4a80-cc35-11ea-9f28-801725a19a3c.PNG">

**Comment by Joanes04** [comment]
Hi @Yashvendra ,

I'm sorry for the previous answer. I just realized that you had selected in the API configuration the `config.basic_auth = "no"`, so the APP should not ask you for those fields defined in the host. 

We have opened an issue #2397 to fix this problem as soon as possible. Thank you very much for reporting this problem.

On the other hand, the new bug seems to have a problem with the versions. It seems that you have the API in a 3.13.1 version, check that the Wazuh UI you have installed corresponds to that version of Wazuh, as it seems that you have installed a 3.12.x version. 

Here is the matrix of compatible versions (https://github.com/wazuh/wazuh-kibana-app#wazuh---kibana---open-distro-version-compatibility-matrix) . You need to install the latest release https://packages.wazuh.com/wazuhapp/wazuhapp-3.13.1_7.8.0.zip. Remember that this last release is for Kibana 7.8.0

Regards,
Alberto

**Comment by Yashvendra** [comment]
Hi @Joanes04,

The version of the Wazuh plugin I installed is also compatible according to this matrix. **Apparently it doesn't work**. 
I was following this installation [guide](https://documentation.wazuh.com/3.12/installation-guide/installing-elastic-stack/elastic_server_deb.html), which I think will need a minor update in _Install the Wazuh app plugin for Kibana_ section.

**Steps to Resolve the API issue:**
I removed the currently installed wazuh (3.12.3) plugin.
```
$ cd /usr/share/kibana
$ sudo -u kibana bin/kibana-plugin remove wazuh
```
Then I installed the latest supported wazuh plugin for Kibana=7.7.1
```
$ sudo -u kibana bin/kibana-plugin install https://packages.wazuh.com/wazuhapp/wazuhapp-3.13.0_7.7.1.zip
$ systemctl restart kibana.service
```
The API issue is resolved now. 

Thanks for your help :)
